plugins/cache/bap_cache_gc.ml

```diff
@@ -0,0 +1,154 @@
+    the total size of the cache is Sum(sizes(i)) = 20
+    the PDF is p(i) = {4/20; 6/20; 3/20; 1/20; 6/20}
+    and CDF is F(i) = {4/20; 10/20; 13/20; 14/20; 20/20}
+
+   We don't want to use floating points, there will be too many big and
+   small numbers and overflows and we finally want to get an
+   index. We will use rational numbers, since formulas 4. and 5. have the
+   same denominator (namely T) we can use only numenators.
+
+   On the high-level, we need to generate a random value between 0 and
+   T, and find such k that F(k-1) < S <= F(k), the k-th file will be
+   our candidate for removal. We can repeat sampling until we get |n|
+   files (of course deleting the same file twice won't free twice of
+   its size, so we had to keep in mind which files we already selected
+   and repeat until we get |n| distinct files)
+   Of course, we don't want to have a linear search for intervals, but
+   we can see, that F(i) partitions the set of sizes (0...T) into m-1
+   subsets, so we can represent F as a finite mapping, e.g., with our
+   example,
+
+   [0,3] -> 0
+   [4,9] -> 1
+   [10,12] -> 2
+   [13,13] -> 3
+   [14,19] -> 4
+
+   Since intervals are not intersecting, we don't need to use
+   Interval_map here, we just need to use the common Map from core
+   with the closest_key (`Less_or_equal_to`` function. So once we
+   generated a random size u we call for the closest_key for u and
+   pick the associated value as the index of the file that we will
+   delete. E.g., let's choose randomly a value from the range of
+   0...19, if it in range from 0..3 we will pick the first file, or if
+   it is in range from 4,9, e.g., 5, then closest_key will return 4,1,
+   so we will remove the second file. So we managed to get away from
+   ugly floats and got the desired distribution with no rounding
+   errors.
+
+   Now, after we have selected |n| distinct files we can shuffle them and
+   delete without worrying that some other process already deleted one
+   of those files. All the processes are using the same sequence of
+   pseudorandom files, so they will select approximately equal files
+   for deletion.
+
+   And finally, we don't want to make our recursive selection depend
+   from |n|, so instead of selecting |n| files for removal we will
+   select as many files as we need to remove requested size,
+*)
+
+open Core_kernel
+open Bap.Std
+
+include Self ()
+
+module Cfg = Bap_cache_config
+module CDF = Int.Map
+
+type entry = {
+  path : string;
+  size : int;    (* Kb  *)
+}
+
+let (//) = Filename.concat
+
+let min_entry_size = 4 (* Kb *)
+
+let entry path name =
+  try
+    let path = path // name in
+    let size = Unix.( (stat path).st_size ) / 1024 in
+    Some {path; size;}
+  with _ -> None
+
+let read_cache path =
+  Sys.readdir path |> Array.filter_map ~f:(entry path)
+
+let total_size =
+  Array.fold ~init:0 ~f:(fun s e -> s + e.size)
+
+let cdf entries =
+  let prob e = max min_entry_size e.size in
```

Comment:
please, let's not call it a probability, just do `pre + max min_entry_size size`